lyse/__init__.py

```diff
@@ -418,32 +578,42 @@ def globals_diff(self, other_run, group=None):
         
 class Sequence(Run):
     def __init__(self, h5_path, run_paths, no_write=False):
```

Comment:
A significant amount of this `__init__` method appears to overlap with the base class. There are minor differences (like `f_locals` vs `f_globals` but I'm not sure there are material differences. I also suspect that making the instance attributes private (see other comment in this review) may prevent the properties of the base class from working in the child class (could be wrong about this). However, calling `super().__init__` in the child class and having the parent set those values would negate the issue (should it exist).

Can you see any reason why `Sequence.__init__` isn't just an extension of `Run.__init__` provided you attempt to create the file if it doesn't exist before calling the parent method?

Reply:
Yeah, I agree that refactoring the `__init__()` methods would have been nice. It's possible but I couldn't get it to be pretty so I abandoned the idea and went with single underscore private attributes. The issue has to do with getting the name of the lyse script.

When calling `Run.__init__()` directly from a lyse script, both `f_globals['__file__']` and `f_locals['__file__']` are the path to the script. However things go wrong when the lyse script calls `Sequence.__init__()` then that calls `Run.__init__()` which then tries to determine its calling script's name. In that case `f_globals['__file__']` gets set to the path to lyse's `__init__.py` and `f_locals` doesn't have a `__file__` key. That meant either `self.group` would be set to `__init__` if using `f_globals` or using `f_locals` would raise a `KeyError` which would then be suppressed.

As a workaround for that I added an optional `group` argument to `Run.__init__()` which defaulted to `None` and would be automatically determined from `f_globals` if it was `None`, so directly calling `Run.__init__()` worked as before. I then had `Sequence.__init__()`  evaluate `f_globals['__file__']` itself, which gave the name of the lyse script, and pass that value to `Run.__init__()`.

That mostly worked, but creating a `Sequence` from outside of lyse had issues. In that case `Sequence.__init__()` would correctly fail to get a name of a lyse script and then would pass `group=None` to `Run.__init__()`. However that would make `Run.__init__()` try to automatically determine the calling script, which it would think was lyse's `__init__.py`, so it would set `self.group` to `__init__` instead of `None`.

I can think of two ways to work around that last issue and make instantiating a `Sequence` from an interpreter work. One option is to have `Sequence.__init__()` pass some other special non-default value for `group` to `Run.__init__()` which would tell it not to try to automatically determine `self.group` and instead just set it to `None`. Another option is to have `Run.__init__()` use `f_locals` instead of `f_globals` so that it fails to automatically get a script name when `Sequence.__init__()` calls it with `group=None`. I can't tell if that's hacky or clever. It definitely makes the code hard to understand though since you have know about some intricacies of how the values in `f_locals` depend on the caller.

Given that the first workaround didn't seem appealing I ended up sticking with the lyse's previous approach with duplicated code and used single underscore attributes. I didn't think of the second workaround until writing this, but I'm not sold on it either to be honest.  Maybe there's a better option?

Reply:
Ah I see. I'll see if I can determine a better workaround!

Reply:
Ok, so I think that the solution might be to pass in `frame=inspect.currentframe().f_back` as an optional argument to `Run`. I suspect that the issue is this (inside `Run.__init__` points to the `Sequence.__init__` method and you would need an additional `f_back` chained on to get out of `Sequence` and back into the lyse script). So obtaining that value from within `Sequence` and passing it to `Run` should resolve the issue I think.

I might be wrong though, because that would imply using a `Run` object from within an imported function would result in the wrong group name being set. But maybe no one has ever tried instantiating a `Run` from within an imported module? If that does result in a bug, then we actually need to inspect the whole stack until we get up to `os.path.join(lyse.LYSE_DIR,'analysis_subprocess.py')` and then go down one level to get the lyse script name (I think?). This solution (if it works) would also solve the problem and would not need the additional kwarg (since it would walk to stack until it finds the right value).

Reply:
Nice! That second idea worked great and is implemented in 3d38b3b.

Reply:
Great! Last nitpick: does your current implementation fail if the analysis script itself is also called analysis_subprocess.py?

Reply:
Yeah I believe it would. I'll adjust it to use your `os.path.join....` suggestion to be safe.

Reply:
Updated in 1dac68f.